### ISSUE TITLE
ManagedSeed sets `debugging.*=false`

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#4654]
**How to categorize this issue?**

/area dev-productivity ops-productivity
/kind bug

**What happened**:

When creating a simple `ManagedSeed` object like this
```yaml
apiVersion: seedmanagement.gardener.cloud/v1alpha1
kind: ManagedSeed
metadata:
  name: gcp
  namespace: garden # Must be garden
spec:
  shoot:
    name: seed-gcp
  gardenlet: {}
```

the newly added `debugging` fields (see https://github.com/gardener/gardener/issues/4567, https://github.com/gardener/gardener/pull/4568) in the gardenlet config will be set to `false,false`:
```yaml
apiVersion: seedmanagement.gardener.cloud/v1alpha1
kind: ManagedSeed
metadata:
  name: gcp
  namespace: garden
spec:
  gardenlet:
    config:
      apiVersion: gardenlet.config.gardener.cloud/v1alpha1
      debugging:
        enableContentionProfiling: false
        enableProfiling: false
      kind: GardenletConfiguration
# ...
    mergeWithParent: true
  shoot:
    name: seed-gcp
```

Because of this, the gardenlet that is deployed for the `ManagedSeed` will be configured with profiling handlers disabled even though the parent gardenlet has profiling handlers enabled and `mergeWithParent=true`.
Config of parent gardenlet:
```yaml
debugging:
  enableProfiling: true
  enableContentionProfiling: true
```

Config of `ManagedSeed` gardenlet:
```yaml
debugging:
  enableProfiling: false
  enableContentionProfiling: false
```

**What you expected to happen**:

I expected that if I create a `ManagedSeed` object without setting the `debugging` fields, the settings in the parent gardenlet are propagated down to the child gardenlet.

**Environment**:

- Gardener version: `v1.31.0`

**Comment by timebertt** [comment]
In the versioned gardenlet config API, `debugging.*` are pointers:
https://github.com/gardener/gardener/blob/862c77153f92be4f9ab8d039d2751f630539ef90/pkg/gardenlet/apis/config/v1alpha1/types.go#L66-L67
https://github.com/gardener/gardener/blob/06242c99ecebf7442021acbb14a895eca27f7f78/vendor/k8s.io/component-base/config/v1alpha1/types.go#L60-L66

However, in the internal gardenlet config API, `debugging.*` are `bool`s:
https://github.com/gardener/gardener/blob/862c77153f92be4f9ab8d039d2751f630539ef90/pkg/gardenlet/apis/config/types.go#L55-L56
https://github.com/gardener/gardener/blob/6806495376b1de8b760b10cf5bc5cf9445f30b06/vendor/k8s.io/component-base/config/types.go#L74-L80

This API design brings up problems when converting `spec.gardenlet.config` back and forth in the `ManagedSeed` admission plugin (I missed this scenario in https://github.com/gardener/gardener/pull/4568#discussion_r701893959)
When converting to internal [here](https://github.com/gardener/gardener/blob/e8eb807d790ff683e650b5918ddf64879863856a/plugin/pkg/managedseed/validator/admission.go#L245) the `nil` bool pointers are converted to `false`.
When converting back to `v1alpha` [here](https://github.com/gardener/gardener/blob/e8eb807d790ff683e650b5918ddf64879863856a/plugin/pkg/managedseed/validator/admission.go#L262), the `false` values are converted to `*false`, which will then be persisted in the store in the resulting `ManagedSeed` object.

To summarize: the problem is that the internal config API doesn't have pointers and that the API server touches the config in its admission plugins.

**Comment by timebertt** [comment]
> To summarize: the problem is that the internal config API doesn't have pointers and that the API server touches the config in its admission plugins.

I see two options for resolving out this situation:
- either we fix it via the API
  - add pointers to `debugging` in the internal and external gardenlet config API
  - ensure proper conversion between the API versions
  - as discussed in https://github.com/gardener/gardener/pull/4568#discussion_r701973036, we are lacking a proper guideline / general approach on such decisions, so we missed this point (but this is probably a good lesson to add to such a guide)
- or we fix it in the `ManagedSeed` admission plugin (like in this WIP: 309549866bab9390fa547833591f8c56f516bf3d)
  - one could argue that this is the conceptual mistake: that the API server touches the gardenlet config at all
  - we are only interested in the `seedConfig` in the internal version
  - so we can only do a back and forth conversion for that part only, but not for the entire gardenlet config
  - this results in `debugging: {}` when leaving out the fields on creation, which results in the expected behavior (don't know why it's not left out though)

I'm looking for opinions here :)

**Comment by stoyanr** [comment]
I would vote for the first option (fix via the API). From my perspective the fact that converting between internal and external versions results in different objects is already a problem that should be avoided. I think we would have it with other types from `component-base/config` as well (e.g. `LeaderElectionConfiguration`) except that they are always used as pointers. Also, `Debugging` is currently the only optional field that is not a pointer (and not marked as optional) in the gardenlet config, so also from consistency reasons it should be changed. I proposed this already in https://github.com/gardener/gardener/pull/4568#discussion_r701893959 but didn't insist on it since I was of course also not able to foresee this particular issue.

I don't have any particular issue with the second proposal, I guess it would work as well. I don't think however that the current approach is conceptually wrong - it converts an external type to an internal one and then back to external, an operation that one would normally expect to result in identical objects (except for the changed introduced by the plugin itself). So I see the proposed fix more as a workaround - although of course an acceptable one if there is a strong reason to avoid the first option (but I don't see one).

**Comment by timebertt** [comment]
Thanks for your input, @stoyanr.

> I don't think however that the current approach is conceptually wrong - it converts an external type to an internal one and then back to external, an operation that one would normally expect to result in identical objects (except for the changed introduced by the plugin itself). So I see the proposed fix more as a workaround

Yeah, I get your point.
Then, let me correct my point from above: probably the conceptual problem is rather, that these component config APIs were never thought to be persisted on the API server (where they are converted back and forth) or included as `RawExtension`s in other objects.
So we should improve how we include them in our gardenlet config API that ens up in the `ManagedSeed` and add a pointer (option 1).
I agree with you that option 2 is actually rather a workaround.

I will proceed with option 1. Thanks again for your opinion :)
